File: shell/android/platform_service_provider.cc
// Copyright 2015 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "sky/shell/android/platform_service_provider.h"

#include "base/android/jni_android.h"
#include "base/bind.h"
#include "base/trace_event/trace_event.h"
#include "jni/PlatformServiceProvider_jni.h"
#include "mojo/public/cpp/bindings/interface_request.h"
#include "sky/shell/service_provider.h"

namespace sky {
namespace shell {
namespace {

void CreatePlatformServiceProvider(
    mojo::InterfaceRequest<mojo::ServiceProvider> request) {
  Java_PlatformServiceProvider_create(
      base::android::AttachCurrentThread(),
      base::android::GetApplicationContext(),
      request.PassMessagePipe().release().value());
}

} // namespace

bool RegisterPlatformServiceProvider(JNIEnv* env) {
  return RegisterNativesImpl(env);
}

mojo::ServiceProviderPtr CreateServiceProvider(
        ServiceProviderContext* context) {
  mojo::MessagePipe pipe;
  context->java_task_runner->PostTask(
      FROM_HERE,
      base::Bind(CreatePlatformServiceProvider,
                 base::Passed(mojo::MakeRequest<mojo::ServiceProvider>(
                     pipe.handle1.Pass()))));
  return mojo::MakeProxy(
      mojo::InterfacePtrInfo<mojo::ServiceProvider>(pipe.handle0.Pass(), 0u));
}

}  // namespace shell
}  // namespace sky

File: sky/shell/platform/android/org/domokit/sky/shell/PlatformServiceProvider.java
// Copyright 2015 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package org.domokit.sky.shell;

import android.content.Context;

import org.chromium.mojo.system.Core;
import org.chromium.mojo.system.MessagePipeHandle;
import org.chromium.mojo.system.MojoException;
import org.chromium.mojo.system.impl.CoreImpl;
import org.chromium.mojom.mojo.ServiceProvider;

/**
 * A collection of services implemented in Java.
 **/
public class PlatformServiceProvider implements ServiceProvider {
    private Core mCore;
    private Context mContext;
    private ServiceRegistry mLocalRegistry;

    public PlatformServiceProvider(Core core, Context context, ServiceRegistry localRegistry) {
        assert core != null;
        assert context != null;
        mCore = core;
        mContext = context;
        mLocalRegistry = localRegistry;
    }

    @Override
    public void close() {}

    @Override
    public void onConnectionError(MojoException e) {}

    @Override
    public void connectToService(String interfaceName, MessagePipeHandle pipe) {
        ServiceFactory factory = mLocalRegistry.get(interfaceName);
        if (factory == null) {
            factory = ServiceRegistry.SHARED.get(interfaceName);
        }
        if (factory == null) {
            pipe.close();
            return;
        }
        factory.connectToService(mContext, mCore, pipe);
    }
}
